subprojects/core/src/main/java/org/gradle/internal/cleanup/DefaultBuildOutputCleanupRegistry.java

```diff
@@ -56,21 +56,28 @@ public boolean isOutputOwnedByBuild(File file) {
     }
 
     private Set<String> getResolvedPaths() {
-        if (resolvedPaths == null) {
-            doResolvePaths();
+        // Use double checked locking,
+        // since `isOutputOwnedByBuild` can be accessed from multiple projects at the same time,
+        // while other projects are still running configuration.
+        ImmutableSet<String> tmp = resolvedPaths;
+        if (tmp == null) {
+            synchronized (this) {
+                tmp = resolvedPaths;
+                if (tmp == null) {
+                    resolvedPaths = tmp = doResolvePaths();
+                }
+            }
         }
-        return resolvedPaths;
+        return tmp;
```

Comment:
Why not do the locking inside `doResolvePaths()`?

Reply:
Because now the whole locking logic is in this method, while the logic how to calculate the paths is in `doResolvePaths`. What would be the advantage of doing the locking in `doResolvePaths()`?